### ISSUE TITLE
Allow creation of a new token even if an already still valid token exists

### ISSUE THREAD

**Issue by gorghoa** [coopTilleuls/CoopTilleulsForgotPasswordBundle#37]
Hi :)

Currently (unless I missed a config option), we can’t ask for a new token if a token has already been asked and still valid regarding its age.

I wonder why this restriction.

As a user, I would rather (and I think it’s what the majority of websites do) be able to ask for a new token even if one exists (invaliding the former one in the process).

The most evident use case I’m thinking of is the email with the token being never received.

If it’s for security reason because hashing is cpu consuming, we could set a delay minimum of a few seconds that would be sufficient in this regard I reckon.

**Comment by gorghoa** [comment]
PS. I may have some time to PR this next week if you think it’s okay to do so.

**Comment by vincentchalamon** [comment]
This restriction is a security to prevent some hack, this kind of security is also present in many websites.
You can change the token lifetime using [`expires_in` configuration option](https://github.com/coopTilleuls/CoopTilleulsForgotPasswordBundle/blob/master/Resources/doc/getting_started.md#configure-your-application):
```yaml
coop_tilleuls_forgot_password:
    password_token:
        expires_in: 1 minute
```

**Comment by gorghoa** [comment]
Thanks @vincentchalamon, I was aware of this config option, but as the doc propose it to 1 day, I thought it was a sensible default!

So: you do recommend to lower it to few minutes?

If you can spare the time, do you have some keywords, links about the kinds of hacks this should solve?

**Comment by vincentchalamon** [comment]
Here is a good article about best practices & risks in a forgot password: https://postmarkapp.com/guides/password-reset-email-best-practices.

In his [example](https://postmarkapp.com/guides/password-reset-email-best-practices#postmarks-password-reset-template), the token duration is 24 hours (like in common websites), that's why I set it as default value in this bundle. But you're free to change this duration in your project, so yeah I recommend you to use this option **for your usage** (I mean not in general).

**Comment by gorghoa** [comment]
https://postmarkapp.com/guides/password-reset-email-best-practices#expiration-information

In fact, I find the 24 hours expiration delay fine. But I nowhere see that the user shouldn’t be able to request a new token within this delay ;)

**Comment by vincentchalamon** [comment]
@gorghoa if you have time, could you please open a PR to improve the README.md file to include this link? It could help people to understand security warnings about this bundle.

**Comment by gorghoa** [comment]
Not until I understand why it’s a security flaw to ask for new token :-p

**Comment by gorghoa** [comment]
https://www.owasp.org/index.php/Forgot_Password_Cheat_Sheet#Step_3.29_Send_a_Token_Over_a_Side-Channel does not say we should not either

**Comment by gorghoa** [comment]
some more reading: https://ux.stackexchange.com/questions/38612/expected-behaviour-when-clicking-to-reset-a-password-multiple-times

(It may seems unfair to only link you to stuffs validating my point of view, but I don’t find the other way)

**Comment by gorghoa** [comment]
(Note that until you don’t tell me to f**k off I’ll continue to argue, I’ll don’t take it personnaly so feel free! :D)

**Comment by vincentchalamon** [comment]
Those articles contain good points that could be explored in different issues/PR.

>If a link is still active, and the 'forgot password'-button is clicked again, resend the active reset link.

Could be a good idea, but I have some doubts about generating a new token and keep the old ones active...

>make sure you limit the number of guesses attempted and if some threshold is exceeded for that user (say 3 to 5), lock out the user's account for some reasonable duration (say at least 5 minutes) and then challenge the user with some form of challenge token per standard multi-factor workflow.

That's a very good idea, but it needs to implement a lot of features: connect with a multi-factor app, lock user account after X attempts, lock IP in case of too many requests... Many of those points are out of this bundle scope imo.

>It is not unreasonable to think that a user's email account may have already been compromised, so tokens that do not involve email, such as SMS or a mobile soft-token, are best.

That's the best part of this comment, but hard to implement (I guess): it could create a dependency to an external service, but could also improve this bundle security.

>It is also a good idea to have the random code which your system generates to only have a limited validity period, say no more than 20 minutes or so.

That's your point @gorghoa, maybe default value should be 20 minutes instead of 24 hours. The main difficulty is to find a correct default time to allow users with low connection to get email after minutes (or more?), and prevent hackers to track this email and connect before the real user has time to reset its password (so the token won't be valid anymore and the hacker will yell of rage :p)

**Comment by gorghoa** [comment]
My point is mainly part of this one (the active delay is already configurable and I’m very comfortable with how it is today):

> If a link is still active, and the 'forgot password'-button is clicked again, resend the active reset link.

except I would be more inclined to send a fresh new token instead of the same one. (The first sent may have been compromised in the delay?)

**Comment by gorghoa** [comment]
aside note:

> It is not unreasonable to think that a user's email account may have already been compromised, so tokens that do not involve email, such as SMS or a mobile soft-token, are best.

I was thinking about that, maybe the easiest way would be to to inject the reset token channel service (defaulting to the built-in provided existing email channel). Then user’s can plug any channel by DI. (it may already be possible though, haven’t checked)

**Comment by vincentchalamon** [comment]
>except I would be more inclined to send a fresh new token instead of the same one. (The first sent may have been compromised in the delay?)

But if the email has been compromised, sending a new one doesn't solve this issue cause any hacker could get this new token from the compromised email (for instance by asking a new token) and hack the user account easily. Sending a unique token during a lifetime prevent that (not at 100%) in case the user uses this token before the hacker does, the token will be invalid.

>(...)it may already be possible though, haven’t checked

It's already possible. On request token generation, the developer should (must) [create a listener to send an email](https://github.com/coopTilleuls/CoopTilleulsForgotPasswordBundle/blob/master/Resources/doc/usage.md#send-email-on-user-request) to the user containing the token. In this point, it's possible to use a SMS service, or whatever.

**Comment by gorghoa** [comment]
Fair enough, we do agree that we can resend an email with token while a token is still valid, so ?

**Comment by vincentchalamon** [comment]
OK for that, but I disagree about generating a new token if one is still valid

**Comment by vincentchalamon** [comment]
@gorghoa Do you also think _lock token generation for user (user account too?) on too many attempts_ should be a good feature in this bundle too? Could increase its security, but I'm afraid it could be out of scope, what do you think?

**Comment by gorghoa** [comment]
Since we dont regenerate a new token while one is still valid, it has few interest imo.

Still, we could temper it by exponentially refuse clicking “send me my token” by a certain delay (maybe by storing the timestamps of times  the token has been sent)

**Comment by vincentchalamon** [comment]
Could be interesting...in another issue :) Feel free to open
